src/test/java/com/google/devtools/build/lib/skyframe/TrackSourceDirectoriesIntegrationTest.java

```diff
@@ -0,0 +1,93 @@
+package com.google.devtools.build.lib.skyframe;
+
+import static com.google.common.truth.Truth.assertThat;
+import static com.google.devtools.build.lib.testutil.MoreAsserts.assertContainsEvent;
+import static com.google.devtools.build.lib.testutil.MoreAsserts.assertDoesNotContainEvent;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.devtools.build.lib.buildtool.util.BuildIntegrationTestCase;
+import com.google.devtools.build.lib.events.EventKind;
+import com.google.devtools.build.lib.events.util.EventCollectionApparatus;
+import com.google.devtools.build.lib.vfs.Path;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link TrackSourceDirectoriesFlag}. */
+@RunWith(JUnit4.class)
+public final class TrackSourceDirectoriesIntegrationTest extends BuildIntegrationTestCase {
+
+  @Override
+  protected EventCollectionApparatus createEvents() {
+    return new EventCollectionApparatus(ImmutableSet.of(EventKind.FINISH));
+  }
+
+  @Before
+  public void setupAndBuildTestTarget() throws Exception {
+    // Enabled via jvm_flags.
+    assertThat(TrackSourceDirectoriesFlag.trackSourceDirectories()).isTrue();
+
+    getWorkspace().getRelative("pkg/dir/empty_dir").createDirectoryAndParents();
+    write("pkg/dir/file", "foo");
+    write("pkg/BUILD",
+        "genrule(",
+        "    name = 'a',",
+        "    srcs = ['dir'],",
+        "    outs = ['out'],",
+        "    cmd = 'touch $@',",
+        ")");
+
+    // Build the target once - all tests below modify the source directory and assert that this
+    // causes the target to be rebuilt.
+    assertTargetRebuilds();
+  }
+
+  @Test
+  public void noChanges_targetDoesNotRebuild() throws Exception {
+    assertTargetDoesNotRebuild();
+  }
+
+  @Test
+  public void fileCreatedInEmptyDir_targetDoesRebuild() throws Exception {
+    write("pkg/dir/empty_dir/file");
+    assertTargetRebuilds();
+  }
+
+  @Test
+  public void fileContentChanged_targetDoesRebuild() throws Exception {
+    write("pkg/dir/file", "changed");
+    assertTargetRebuilds();
+  }
+
+  @Test
+  public void emptyDirectoryCreated_targetDoesRebuild() throws Exception {
+    getWorkspace().getRelative("pkg/dir/empty_dir/nested_empty_dir").createDirectory();
+    assertTargetRebuilds();
+  }
+
+  @Test
+  public void emptyDirectoryRemoved_targetDoesRebuild() throws Exception {
+    getWorkspace().getRelative("pkg/dir/empty_dir").deleteTree();
+    assertTargetRebuilds();
+  }
+
+  @Test
+  public void emptyDirectoryReplacedWithEmptyFile_targetDoesRebuild() throws Exception {
+    getWorkspace().getRelative("pkg/dir/empty_dir").deleteTree();
+    write("pkg/dir/empty_dir");
+    assertTargetRebuilds();
+  }
+
+  private void assertTargetRebuilds() throws Exception {
+    events.collector().clear();
```

Comment:
extreme-nitpick: You could consider moving clear to the setup function and if you prefer, maybe even drop these helpers altogether, again, up to you.